vms/platformvm/state/chain_time_helpers.go

```diff
@@ -64,42 +75,83 @@ func GetNextStakerChangeTime(state Chain, defaultTime time.Time) (time.Time, err
 		}
 
 		time := it.Value().NextTime
-		if time.Before(defaultTime) {
-			defaultTime = time
+		if time.Before(nextTime) {
+			nextTime = time
 		}
 	}
-	return defaultTime, nil
+
+	sovIterator, err := state.GetActiveSubnetOnlyValidatorsIterator()
+	if err != nil {
+		return time.Time{}, err
+	}
+	defer sovIterator.Release()
+
+	// If there are no SoVs, return
+	if !sovIterator.Next() {
+		return nextTime, nil
+	}
+
+	// Calculate the remaining funds that the next validator to evict has.
+	var (
+		sov         = sovIterator.Value()
+		accruedFees = state.GetAccruedFees()
+	)
+	remainingFunds, err := math.Sub(sov.EndAccumulatedFee, accruedFees)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	// Calculate how many seconds the remaining funds can last for.
+	var (
+		currentTime = state.GetTimestamp()
+		maxSeconds  = uint64(nextTime.Sub(currentTime) / time.Second)
+	)
+	feeState := validatorfee.State{
+		Current: gas.Gas(state.NumActiveSubnetOnlyValidators()),
+		Excess:  state.GetSoVExcess(),
+	}
+	remainingSeconds := feeState.SecondsRemaining(
+		config,
+		maxSeconds,
+		remainingFunds,
+	)
+
+	deactivationTime := currentTime.Add(time.Duration(remainingSeconds) * time.Second)
+	if deactivationTime.Before(nextTime) {
+		nextTime = deactivationTime
+	}
+	return nextTime, nil
 }
 
 // PickFeeCalculator creates either a static or a dynamic fee calculator,
 // depending on the active upgrade.
 //
 // PickFeeCalculator does not modify [state].
-func PickFeeCalculator(cfg *config.Config, state Chain) fee.Calculator {
+func PickFeeCalculator(config *config.Config, state Chain) txfee.Calculator {
```

Comment:
`cfg` was changed to `config` to keep consistency with the modified function above (and the rest of the repo)